spock-core/src/main/java/org/spockframework/runtime/SafeIterationNameProvider.java

```diff
@@ -17,20 +17,52 @@
 import org.spockframework.runtime.model.IterationInfo;
 import org.spockframework.runtime.model.NameProvider;
 
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
 public class SafeIterationNameProvider implements NameProvider<IterationInfo> {
   private final NameProvider<IterationInfo> delegate;
-  private int iterationCount;
-  
+  private final AtomicInteger iterationCount = new AtomicInteger(0);
+  private final ConcurrentMap<String, Boolean> usedNames = new ConcurrentHashMap<String, Boolean>();
+
   public SafeIterationNameProvider(NameProvider<IterationInfo> delegate) {
     this.delegate = delegate;
   }
-  
+
   public String getName(IterationInfo iteration) {
-    String safeName = iteration.getParent().isReportIterations() ? 
-        String.format("%s[%d]", iteration.getParent().getName(), iterationCount++) : iteration.getParent().getName();
-    
+    final String name = getNameImpl(iteration);
+    if (iteration.getParent().isReportIterations()){
+      return unique(name);
+    }else {
+      return name;
+    }
+  }
+
+    private String unique(String name) {
```

Comment:
If you use @Unroll annotation with template - you will provide your own names, not based on iteration number.